### DISCUSSION TITLE
find_geom for Model Editing Python API to update size dynamically

### DISCUSSION THREAD

**Discussion by curtiscjohnson** [google-deepmind/mujoco#2295] [Asking for Help] [answered]
### Intro

Hi!

I am a PhD student at BYU, I use MuJoCo for my research on robotic manipulation.

### My setup

MuJoCo v3.2.3, Ubuntu 20.04, Python API

### My question

I've been working on implementing methods to dynamically update an object's size (i.e. in between episodes of reinforcement learning training). I've had lots of odd contact behavior attempting to dynamically update size. Below is a MWE of what I was attempting to do--and a video of the resulting behavior: 

https://github.com/user-attachments/assets/e5958470-6bd4-42f7-be4c-e1110f3ea5b0

I expected the box to just sit on the ground plane, but instead if penetrates and bounces around randomly. _Why does this happen? What am I missing?_

I've read through related Github issues (e.g. #453) and it seems that the recommended way to approach this is the new [Model Editing API](https://mujoco.readthedocs.io/en/stable/programming/modeledit.html#model-editing) via mjSpec. That's great, as it does seem to solve the above bouncing issues. The box size changes correctly and the box just sits on the ground plane.

 MWE with mjSpec:
```python
import mujoco as mj
import mujoco.viewer as viewer
import time

xml_string = """
<?xml version="1.0"?>
<mujoco model="box_on_ground">

    <!-- Worldbody -->
    <worldbody>
        <!-- Ground plane -->
        <geom name="ground" type="plane" pos="0 0 0" size="5 5 0.1" rgba="0.7 0.7 0.7 1" />

        <!-- Box -->
        <body name="box" pos="0 0 0">
            <geom name="box_geom" type="box" size="0.05 0.05 0.05" rgba="1 0 0 1" />
            <freejoint name="free" />
        </body>
    </worldbody>


</mujoco>
"""

mjspec = mj.MjSpec()
mjspec.from_string(xml_string)

box = mjspec.find_body("box")

box_geom = box.first_geom()
box_geom.size = [1 / 2, 1 / 2, 1 / 2]

model = mjspec.compile()
data = mj.MjData(model)

#still place object on the ground
data.joint("free").qpos[:3] = [0, 0, 1 / 2]

with viewer.launch_passive(model, data) as viewer:
    viewer.sync()
    input()
    while viewer.is_running():
        step_start = time.time()
        mj.mj_step(model, data)
        viewer.sync()

        time_until_next_step = model.opt.timestep - (time.time() - step_start)
        if time_until_next_step > 0:
            time.sleep(time_until_next_step)
```

The only way I could figure out how to access the geom to change its size was via ```first_geom()``` or with Mujoco 3.2.6 the ```geoms = spec.worldbody.find_all(mj.mjtObj.mjOBJ_GEOM)``` function from the [Python Model Editing tutorial notebook](https://colab.research.google.com/github/google-deepmind/mujoco/blob/main/python/mjspec.ipynb). 

My question here is this: _Is there a more direct way of accessing geoms through the model editing API that avoids looping through items using ```first_geom()``` and ```next_geom()``` or ```find_all()``` if there are several geoms attached to the same body?_ For example, a ```find_geom("box_geom")``` function?

### Minimal model and/or code that explain my question

```python
import mujoco as mj
import mujoco.viewer as viewer
import time

xml_string = """
<mujoco model="box_on_ground">

    <!-- Worldbody -->
    <worldbody>
        <!-- Ground plane -->
        <geom name="ground" type="plane" pos="0 0 0" size="5 5 0.1" rgba="0.7 0.7 0.7 1" />

        <!-- Box -->
        <body name="box" pos="0 0 0">
            <geom name="box_geom" type="box" size="0.05 0.05 0.05" rgba="1 0 0 1" />
            <freejoint name="free" />
        </body>
    </worldbody>


</mujoco>
"""

model = mj.MjModel.from_xml_string(xml_string)
data = mj.MjData(model)

#update box geom size
model.geom("box_geom").size = [1 / 2, 1 / 2, 1 / 2]

#update joint position in qpos to set box on ground
data.joint("free").qpos[:3] = [0, 0, 1 / 2]

with viewer.launch_passive(model, data) as viewer:
    viewer.sync()
    input()
    while viewer.is_running():
        step_start = time.time()
        mj.mj_step(model, data)
        viewer.sync()

        time_until_next_step = model.opt.timestep - (time.time() - step_start)
        if time_until_next_step > 0:
            time.sleep(time_until_next_step)
```

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment]
There is `body.geoms`, as documented [here](https://mujoco.readthedocs.io/en/latest/python.html#tree-traversal)

Is that what you're looking for?

**Reply by curtiscjohnson** [reply]
Not quite, as I'd still need to iterate through the list returned by ```body.geoms``` to find the geom I'm looking for.

**Comment by yuvaltassa** [comment, marked as answer]
And of course if you know the name there is always [mjs_findElement](https://mujoco.readthedocs.io/en/latest/APIreference/APIfunctions.html#mjs-findelement): `mjs_findElement(spec, mujoco.mjtObj.mjOBJ_GEOM, 'my_geom')`

I **think** you can also use a string here like `mjs_findElement(spec, 'geom', 'my_geom')`, but I'm not sure if we implemented that 🙂

**Reply by curtiscjohnson** [reply]
Yes, I think this is what I was looking for. But has ```mjs_findElement()``` been wrapped for the Python API? I can't seem to find any of the ```mjs``` functions in Python. Is this usage correct?

```python

import mujoco

print(mujoco.__version__)

mujoco.mjs_findElement()
```

returns

```bash
3.2.6
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-8-99ff7842dbd0>](https://localhost:8080/#) in <cell line: 5>()
      3 print(mj.__version__)
      4 
----> 5 mujoco.mjs_findElement()
      6 

AttributeError: module 'mujoco' has no attribute 'mjs_findElement'
```

**Reply by yuvaltassa** [reply]
Ya you're right. @quagla is working on improving the Python find functionality. Ideally we want something like `spec.find('geom', 'my_geom')` but pybind is being difficult with typing. We might end up just adding all the possible `spec.find_xxx()` functions... TBD. We'll update this issue.

**Reply by curtiscjohnson** [reply]
Okay, great. I'll iterate through using ```body.geoms``` for now until future API updates. Thanks!
